packages/bff/src/services/clientService.ts

```diff
@@ -0,0 +1,358 @@
+        headers,
+      });
+      return { key: key.encodedPem };
+    },
+
+    async createConsumerClient(
+      seed: authorizationApi.ClientSeed,
+      { logger, headers }: WithLogger<BffAppContext>
+    ): Promise<{ id: string }> {
+      logger.info(`Creating consumer client with name ${seed.name}`);
+
+      return authorizationProcessClient.client.createConsumerClient(seed, {
+        headers,
+      });
+    },
+
+    async createApiClient(
+      seed: authorizationApi.ClientSeed,
+      { logger, headers }: WithLogger<BffAppContext>
+    ): Promise<{ id: string }> {
+      logger.info(`Creating api client with name ${seed.name}`);
+
+      return authorizationProcessClient.client.createApiClient(seed, {
+        headers,
+      });
+    },
+  };
+}
+
+export type ClientService = ReturnType<typeof clientServiceBuilder>;
+
+async function enhanceClient(
+  apiClients: PagoPAInteropBeClients,
+  client: authorizationApi.Client,
+  ctx: WithLogger<BffAppContext>
+): Promise<bffApi.Client> {
+  const consumer = await apiClients.tenantProcessClient.tenant.getTenant({
+    params: { id: client.consumerId },
+    headers: ctx.headers,
+  });
+
+  const purposes = await Promise.all(
+    client.purposes.map((p) => enhancePurpose(apiClients, p, ctx))
+  );
+
+  return {
+    id: client.id,
+    name: client.name,
+    description: client.description,
+    kind: client.kind,
+    createdAt: client.createdAt,
+    consumer: {
+      id: consumer.id,
+      name: consumer.name,
+    },
+    purposes,
+  };
+}
+
+async function enhancePurpose(
+  {
+    catalogProcessClient,
+    tenantProcessClient,
+    purposeProcessClient,
+  }: PagoPAInteropBeClients,
+  clientPurposeId: string,
+  { headers }: WithLogger<BffAppContext>
+): Promise<bffApi.ClientPurpose> {
+  const purpose = await purposeProcessClient.getPurpose({
+    params: { id: clientPurposeId },
+    headers,
+  });
+
+  const eservice = await catalogProcessClient.getEServiceById({
+    params: { eServiceId: purpose.eserviceId },
+    headers,
+  });
+
+  const producer = await tenantProcessClient.tenant.getTenant({
+    params: { id: eservice.producerId },
+    headers,
+  });
+
+  return {
+    purposeId: purpose.id,
+    title: purpose.title,
+    eservice: {
+      id: eservice.id,
+      name: eservice.name,
+      producer: {
+        id: producer.id,
+        name: producer.name,
+        kind: producer.kind,
+      },
+    },
+  };
+}
+
+async function getSelfcareUserById(
+  selfcareClient: SelfcareV2UsersClient,
+  userId: string,
+  selfcareId: string
+) {
+  try {
+    return selfcareClient.getUserInfoUsingGET({
+      params: { id: userId },
+      queries: { institutionId: selfcareId },
+    });
+  } catch (error) {
+    throw userNotFound(userId, selfcareId);
+  }
+}
+
+async function decorateKey(
+  selfcareClient: SelfcareV2UsersClient,
+  key: authorizationApi.Key,
+  selfcareId: string
+): Promise<bffApi.PublicKey> {
+  const user = await getSelfcareUserById(
```

Comment:
I see that this method throws a userNotFound, but in the Scala code when calling `getSelfcareUserById` in `decorateKey`and the user cannot be found it does not throw but returns an empty  `UserResponse`,  which seems the reason why then you set some defaults in `toBffApiCompactUser`. So here you should probably catch the error and set the response to {}.

See:
https://github.com/pagopa/interop-be-backend-for-frontend/blob/b99cc2f837c6ef49a2eebf757bb74ecf4bf5e9fc/src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/ClientsApiServiceImpl.scala#L338

This is also consistent with the fact that you don't map a userNotFound error in the error mapper for functions that use `decorateKey`

Reply:
Also, if there is no point where the call to getSelfcareUserById should throw and the error then is handled, you could directly modify getSelfcareUserById to return empty UserResponse instead of throwing

Reply:
Yes that is in fact correct, but i have a few doubts:

1. In that Scala snippet, are all errors catched into an empty user response or only a specific `UserNotFound` error? (i see a `_` pattern match, but i am not sure about that behaviour there)
2. Actually, in the Api converter in Scala, they does set default values: (see https://github.com/pagopa/interop-be-backend-for-frontend/blob/b99cc2f837c6ef49a2eebf757bb74ecf4bf5e9fc/src/main/scala/it/pagopa/interop/backendforfrontend/service/types/SelfcareV2ClientServiceTypes.scala#L29)

Reply:
In Scala it seems to only recover from UserNotFound error, and that error is thrown when the `getUserInfoUsingGet` returns 404: https://github.com/pagopa/interop-be-backend-for-frontend/blob/b99cc2f837c6ef49a2eebf757bb74ecf4bf5e9fc/src/main/scala/it/pagopa/interop/backendforfrontend/service/impl/SelfcareV2ClientServiceImpl.scala#L99

But, I think we could also simplify and return an empty UserResponse whenever we cannot retrieve a user when calling the Selfcare client getUserInfo? Let's also see what other reviewers say :)

Reply:
I'm removing the change request to avoid blocking the PR while I'm out of the office, remember to address this anyways before merging :)